### PR TITLE
return the original status code set by mutating webhook admission

### PR DESCRIPTION
**Pull request by CaoShuFeng**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
/assign @sttts   
Thanks.
**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 1]
```diff
@@ -69,6 +69,9 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr *generic.Version
 			}
 			glog.Warningf("Failed calling webhook, failing closed %v: %v", hook.Name, err)
 		}
+		if statusError, ok := err.(*apierrors.StatusError); ok {
```
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: This makes sure the remote http status code could be sent to the client.

https://github.com/kubernetes/kubernetes/pull/64827/files#diff-e057ec5b2ec27b4ba1e1a3915f715262R100

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 2]
```diff
@@ -69,6 +69,9 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr *generic.Version
...
+			return statusError
+		}
 		return apierrors.NewInternalError(err)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: Do you have any insight about this line? What do we transform into an internal error at all?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: I would expect that, like for the validating dispatcher.

Do we have any e2e / integration coverage for the error message?

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: seems mutatingDispatcher.callAttrMutatingHook() only return two kinds of err: 

1. StatusError

1. ErrCallingWebhook

So I think no need to wrap the err again `apierrors.NewInternalError(err)`, directly return err is ok.

Comment (CaoShuFeng) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: > I would expect that, like for the validating dispatcher.

Done. 

> Do we have any e2e / integration coverage for the error message?

The e2e test check the error message, but doesn't check the http status code.
https://github.com/kubernetes/kubernetes/blob/63c90bb47ec99247cbf29542edfc443ffda7298b/test/e2e/apimachinery/webhook.go#L624

I add status code code check in the unit test.

**Comment by sttts** [comment]
@CaoShuFeng I think this is fine now.

Could you trace back where the code came from? I only did a refactoring of the dispatchers. Would get an ok from the original author.

**Comment by CaoShuFeng** [comment]
> Could you trace back where the code came from? I only did a refactoring of the dispatchers. Would get an ok from the original author.

/assign @caesarxuchao 
Please take a look, thanks.

**Comment by CaoShuFeng** [comment]
Ping @caesarxuchao 
Please take a look about this.

**Review by hzxuzhonghu — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
any update?

**Pull request closed without merging**